File: nodejs/tests/light/apds9960_test.js
'use strict';

const { I2CConnection } = require('../../packages/periph/src/connection/i2c');
const { APDS9960Full }   = require('../../packages/periph/src/chips/light/apds9960');

const I2C_BUS  = parseInt(process.env.I2C_BUS  || '1',  10);
const I2C_ADDR = parseInt(process.env.I2C_ADDR  || '0x39', 16);

let passed = 0;
let failed = 0;

function checkEq(label, got, expected) {
    if (got === expected) {
        console.log('PASS', label);
        passed++;
    } else {
        console.log(`FAIL ${label}: got 0x${got.toString(16).toUpperCase().padStart(2,'0')}, expected 0x${expected.toString(16).toUpperCase().padStart(2,'0')}`);
        failed++;
    }
}

function checkTrue(label, condition) {
    if (condition) { console.log('PASS', label); passed++; }
    else           { console.log('FAIL', label); failed++; }
}

async function main() {
    const connection = new I2CConnection(I2C_BUS, I2C_ADDR);
    const apds = new APDS9960Full(connection);

    checkEq('chip_id', await apds.chipId(), 0xAB);

    const { clear, red, green, blue } = await apds.color();
    checkTrue('color_clear >= 0', clear >= 0);
    checkTrue('color_red >= 0', red >= 0);
    checkTrue('color_green >= 0', green >= 0);
    checkTrue('color_blue >= 0', blue >= 0);

    checkTrue('is_als_valid', await apds.isAlsValid());

    await apds.enableProximity(true);
    const end1 = Date.now() + 100;
    while (Date.now() < end1) {}
    const p = await apds.proximity();
    checkTrue('proximity <= 255', p <= 255);
    checkTrue('is_proximity_valid', await apds.isProximityValid());

    await apds.configureAls(0xB6, 1);
    const end2 = Date.now() + 210;
    while (Date.now() < end2) {}
    checkTrue('als_valid after configure', await apds.isAlsValid());

    await apds.alsThreshold(100, 60000);
    await apds.proximityThreshold(10, 200);
    await apds.setPersistence(0, 1);
    checkTrue('persistence set', true);

    await apds.enableAlsInterrupt(true);
    await apds.enableProximityInterrupt(true);
    await apds.clearAlsInterrupt();
    await apds.clearProximityInterrupt();
    await apds.clearAllInterrupts();
    checkTrue('interrupts cleared', true);

    await apds.setProximityOffset(10, -5);
    await apds.setProximityMask(false, false, false, false);
    checkTrue('proximity offset/mask set', true);

    await apds.enableGesture(true);
    await apds.configureGesture(1, 0, 0, 1, 1, 50, 20);
    checkTrue('gesture configured', true);
    checkTrue('gesture_fifo_level >= 0', (await apds.gestureFifoLevel()) >= 0);
    await apds.clearGestureFifo();
    await apds.enableGestureInterrupt(false);
    await apds.enableGesture(false);
    checkTrue('gesture disabled', true);

    const s = await apds.status();
    checkTrue('status readable', s >= 0);

    await apds.enableProximity(false);

    await connection.close();

    console.log(`===DONE: ${passed} passed, ${failed} failed===`);
    process.exit(failed === 0 ? 0 : 1);
}

main();

File: python/tests/light/apds9960_test.py
import time
import _testconfig as cfg
from periph.transport.i2c_micropython import I2CTransport
from periph.chips.light.apds9960 import APDS9960Full

from machine import I2C, Pin

passed = 0
failed = 0


def check_eq(label, got, expected):
    global passed, failed
    if got == expected:
        print('PASS', label)
        passed += 1
    else:
        print('FAIL {}: got 0x{:02X}, expected 0x{:02X}'.format(label, got, expected))
        failed += 1


def check_true(label, condition):
    global passed, failed
    if condition:
        print('PASS', label)
        passed += 1
    else:
        print('FAIL', label)
        failed += 1


i2c = I2C(cfg.I2C_ID, sda=Pin(cfg.SDA), scl=Pin(cfg.SCL), freq=cfg.FREQ)
transport = I2CTransport(i2c, cfg.ADDR)
apds = APDS9960Full(transport)

check_eq('chip_id', apds.chip_id(), 0xAB)

c, r, g, b = apds.color()
check_true('color_clear >= 0', c >= 0)
check_true('color_red >= 0', r >= 0)
check_true('color_green >= 0', g >= 0)
check_true('color_blue >= 0', b >= 0)

check_true('is_als_valid', apds.is_als_valid())

apds.enable_proximity(True)
time.sleep_ms(100)
p = apds.proximity()
check_true('proximity >= 0', p >= 0)
check_true('proximity <= 255', p <= 255)
check_true('is_proximity_valid', apds.is_proximity_valid())

apds.configure_als(0xB6, 1)
time.sleep_ms(210)
check_true('als_valid after configure', apds.is_als_valid())

apds.als_threshold(100, 60000)
apds.proximity_threshold(10, 200)
apds.set_persistence(0, 1)
check_true('persistence set', True)

apds.enable_als_interrupt(True)
apds.enable_proximity_interrupt(True)
apds.clear_als_interrupt()
apds.clear_proximity_interrupt()
apds.clear_all_interrupts()
check_true('interrupts cleared', True)

apds.set_proximity_offset(10, -5)
apds.set_proximity_mask(False, False, False, False)
check_true('proximity offset/mask set', True)

apds.enable_gesture(True)
apds.configure_gesture(1, 0, 0, 1, 1, 50, 20)
check_true('gesture configured', True)
check_true('gesture_fifo_level >= 0', apds.gesture_fifo_level() >= 0)
apds.clear_gesture_fifo()
apds.enable_gesture_interrupt(False)
apds.enable_gesture(False)
check_true('gesture disabled', True)

s = apds.status()
check_true('status is int', isinstance(s, int))

apds.enable_proximity(False)

print('===DONE: {} passed, {} failed==='.format(passed, failed))
